x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,270 @@
+        }
+
+        if (modelsForField.size() > 1) {
+            // TODO: Handle multi-index semantic queries
+            throw new IllegalArgumentException("Field [" + fieldName + "] has multiple models associated with it");
+        }
+
+        InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+            TaskType.ANY,
+            modelsForField.iterator().next(),
+            List.of(query),
+            Map.of(),
+            InputType.SEARCH
+        );
+
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction((client, listener) -> executeAsyncWithOrigin(
+            client,
+            ML_ORIGIN,
+            InferenceAction.INSTANCE,
+            inferenceRequest,
+            listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                inferenceResultsSupplier.set(inferenceResponse.getResults());
+                l.onResponse(null);
+            })
+        ));
+
+        return new SemanticQueryBuilder(this, inferenceResultsSupplier);
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        InferenceServiceResults inferenceServiceResults = inferenceResultsSupplier.get();
+        if (inferenceServiceResults == null) {
+            throw new IllegalArgumentException("Inference results supplier for field [" + fieldName + "] is empty");
+        }
+
+        List<? extends InferenceResults> inferenceResultsList = inferenceServiceResults.transformToCoordinationFormat();
+        if (inferenceResultsList.isEmpty()) {
+            throw new IllegalArgumentException("No inference results retrieved for field [" + fieldName + "]");
+        } else if (inferenceResultsList.size() > 1) {
+            // TODO: How to handle multiple inference results?
+            throw new IllegalArgumentException(inferenceResultsList.size() + " inference results retrieved for field [" + fieldName + "]");
+        }
+
+        InferenceResults inferenceResults = inferenceResultsList.get(0);
+        MappedFieldType fieldType = context.getFieldType(fieldName);
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType == false) {
+            // TODO: Better exception type to throw here?
+            throw new IllegalArgumentException(
+                "Field [" + fieldName + "] is not registered as a " + SemanticTextFieldMapper.CONTENT_TYPE + " field type"
+            );
+        }
+
+        return semanticQuery(inferenceResults, context);
+    }
+
+    private Query semanticQuery(InferenceResults inferenceResults, SearchExecutionContext context) {
+        // Cant use QueryBuilders.boolQuery() because a mapper is not registered for <field>.inference, causing
+        // TermQueryBuilder#doToQuery to fail (at TermQueryBuilder:202)
+        // TODO: Handle boost and queryName
```

Comment:
[TextExpansionQueryBuilder](https://github.com/elastic/elasticsearch/blob/24d7043828317f8a23b644d04f1e1945c9409e0f/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilder.java#L211) uses QueryBuilders.boolQuery() to build the boolean query, which has the side-effect of handling boost and queryName in [AbstractQueryBuilder#toQuery](https://github.com/elastic/elasticsearch/blob/e4feaff900e259f6accd3ea3756652fe5f0a594d/server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java#L116). However, that doesn't work here because we then have to use [TermQueryBuilder](https://github.com/elastic/elasticsearch/blob/19257125b19ff05a6b014f37693f9417ca579626/server/src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java) to build term queries, which fails in this context in [doToQuery](https://github.com/elastic/elasticsearch/blob/19257125b19ff05a6b014f37693f9417ca579626/server/src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java#L202) because semantic_query relies on "shadow" mappings that are coordinated between the query & [SemanticTextInferenceResultFieldMapper](https://github.com/elastic/elasticsearch/blob/59194d77150702aff723c5202baa4d28f680f2d0/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextInferenceResultFieldMapper.java).

I see two ways to resolve this:

- Duplicate the boost & queryName handling logic here
-  Factor out the boost & queryName handling logic to a static util method and call it here

WDYT?